Workflow Update: update docs with recent code changes (#6506)
temporalio/temporal: docs/architecture/workflow-update.md

> #### TODO
> This might be possible in the future when "Durable Admitted" Update is implemented.

There two Workflow completion scenarios:
- final completion, when no new run is created (e.g. with `COMPLETE_WORKFLOW_EXECUTION` command),
- continued completion, when new run is created (e.g. `CONTINUE_AS_NEW_WORKFLOW_EXECUTION` command).

When the Workflow performs a final completion, all in-flight Updates are aborted: admitted Updates get
`ErrWorkflowCompleted` error on both `accepted` and `completed` futures, and accepted Updates get
the same error on the `completed` future only. It is the Workflow's responsibility to complete
accepted Updates. This behavior is similar to Activities which also can be left abandoned after 
Workflow completed.

> If Workflow completion command creates a new run, accepted Updates are aborted in the same way:
> with `ErrWorkflowCompleted` error on the `completed` future. Admitted Updates, though, are aborted
> with retryable `ErrWorkflowClosing` error. SDK client retries this error and new attempt should land
> on the new run. This is because Updates that were received while the Workflow Task was running, haven't
> been seen by the Workflow yet, and can be safely retried on the new run. It also provides better
> experience for API callers: they shouldn't notice that the Workflow started a new run.
> 
> Aborted accepted (but not completed) Updates remain in the `Aborted` state in the registry even after
> Update registry is reconstructed from the history.

Reviewer: ```suggestion
When a Workflow completion command creates a new run, accepted Updates are aborted in the same way:
with the `ErrWorkflowCompleted` error on the `completed` future. Admitted Updates, though, are aborted
with the retryable `ErrWorkflowClosing` error. The SDK client retries this error and a new attempt should land
on the new run. This is safe because Updates that were received while the Workflow Task was running
haven't been seen by the Workflow yet. It also provides a better experience
for API callers since they shouldn't notice that the Workflow started a new run.

Aborted accepted (but not completed) Updates remain in the `Aborted` state in the registry even after
the Update registry is reconstructed from the history.
```

Update results are available after the Workflow completed. They can be accessed using the
`PollWorkflowExecutionUpdate` API. Note that the `UpdateWorkflowExecution` API will return the
result, too. This provides a consistent experience for API caller: no matter at what stage the 
Workflow is, the API caller will always get the `ErrWorkflowCompleted` error, when the Update wasn't
processed by the Workflow, or the Update outcome when it was.

> #### NOTE

> > If Workflow receives an Update Response message *and* completion command which creates

Reviewer: ```suggestion
> If a Workflow receives an Update Response message *and* a completion command which creates
```

rollback - the transition after a successful persistence write. Check the
[`effect` package doc](./effect-package.md) for more details.

If a Workflow Update is accepted and completed in the same Workflow Task, it goes through the
following chain of state transitions:
```
Sent -> ProvisionalyAccepted -> ProvisionalyCompleted -> ProvisionalyCompletedAfterAccepted -> Completed
```

> Extra `ProvisionalyCompletedAfterAccepted` state is needed to unblock `completed` future before

Reviewer: ```suggestion
The `ProvisionalyCompletedAfterAccepted` in-between state is needed to unblock `completed` future before
```
